### PR TITLE
feat: introduce registries

### PR DESCRIPTION
**Pull request by lsheva**
Inspired by:
https://github.com/rob-Hitchens/UnorderedKeySet
https://medium.com/robhitchens/solidity-crud-epilogue-e563e794fde

- using upsert implementation for register() call to minimize size of smart contract
- using bytes32 insted of uuid string value to reduce memory
- slightly renamed registries function names for brevity and so they all named the same between registries (get, getAll, getByIndex, register, deregister)
- added events
- missing validation on number of tags
- missing mappings for constant time retrieval (`getAgentsByOwner`, `getModelsByOwner`)

### CODE REVIEW

File: smart-contracts/contracts/AgentRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    } 
+    map[uuid] = Agent({
+      fee: fee,
+      stake: amount + stake,
+      timestamp: timestamp,
+      owner: owner,
+      uuid: uuid,
+      name: name,
+      tags: tags
+    });
```
[the remark this answers is not in the document]
Comment (lsheva) [reply smart-contracts/contracts/AgentRegistry.sol:71]: it returns always true, look at the source code

File: smart-contracts/contracts/KeySet.sol [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        if (!exists(self, key)) {
+            revert KeyNotFound();
+        }
+        
+        bytes32 keyToMove = self.keyList[count(self)-1];
+        uint rowToReplace = self.keyPointers[key];
+        self.keyPointers[keyToMove] = rowToReplace;
+        self.keyList[rowToReplace] = keyToMove;
+        delete self.keyPointers[key];
+        self.keyList.pop();
```
[the remark this answers is not in the document]
Comment (lsheva) [reply smart-contracts/contracts/KeySet.sol:37]: one element is not worthy extra branching

File: smart-contracts/contracts/AgentRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  using KeySet for KeySet.Set;
+
+  struct Agent {
+    uint256 fee;
+    uint256 stake;
+    uint256 timestamp;
+    address owner;
+    bytes32 uuid;
+    string name;        // limit name length
+    string[] tags;      // TODO: limit tags amount
```
[the remark this answers is not in the document]
Comment (srt0422) [reply smart-contracts/contracts/AgentRegistry.sol:18]: created an issue to address this later.

File: smart-contracts/contracts/KeySet.sol [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    function count(Set storage self) internal view returns (uint) {
+        return (self.keyList.length);
+    }
+
+    function exists(
+        Set storage self,
+        bytes32 key
+    ) internal view returns (bool) {
+        if (self.keyList.length == 0) return false;
+        return self.keyList[self.keyPointers[key]] == key;
```
[the remark this answers is not in the document]
Comment (srt0422) [reply smart-contracts/contracts/KeySet.sol:49]: created new issue to fix this later

**Review by srt0422 — APPROVED** [review]

**Pull request merged**